cmd/relay-gossip/node/relaynode.go

```diff
@@ -0,0 +1,201 @@
+	"fmt"
+	"time"
+
+	"github.com/drand/drand/cmd/relay-gossip/lp2p"
+	"github.com/drand/drand/protobuf/drand"
+	"github.com/gogo/protobuf/proto"
+	bds "github.com/ipfs/go-ds-badger2"
+	"github.com/libp2p/go-libp2p-core/crypto"
+	"github.com/libp2p/go-libp2p-core/host"
+	pubsub "github.com/libp2p/go-libp2p-pubsub"
+	ma "github.com/multiformats/go-multiaddr"
+	"github.com/prometheus/common/log"
+	"golang.org/x/xerrors"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials"
+)
+
+// GossipRelayConfig configures a gossip relay node.
+type GossipRelayConfig struct {
+	Network         string
+	PeerWith        []string
+	Addr            string
+	DataDir         string
+	IdentityPath    string
+	CertPath        string
+	Insecure        bool
+	DrandPublicGRPC string
+}
+
+// GossipRelayNode is a gossip relay runtime.
+type GossipRelayNode struct {
+	bootstrap []ma.Multiaddr
+	ds        *bds.Datastore
+	priv      crypto.PrivKey
+	h         host.Host
+	ps        *pubsub.PubSub
+	t         *pubsub.Topic
+	opts      []grpc.DialOption
+	addr      []ma.Multiaddr
+	done      chan struct{}
+}
+
+// NewGossipRelayNode starts a new gossip relay node.
+func NewGossipRelayNode(cfg *GossipRelayConfig) (*GossipRelayNode, error) {
+	bootstrap, err := ParseMultiaddrSlice(cfg.PeerWith)
+	if err != nil {
+		return nil, xerrors.Errorf("parsing peer-with: %w", err)
+	}
+
+	ds, err := bds.NewDatastore(cfg.DataDir, nil)
+	if err != nil {
+		return nil, xerrors.Errorf("opening datastore: %w", err)
+	}
+
+	priv, err := lp2p.LoadOrCreatePrivKey(cfg.IdentityPath)
+	if err != nil {
+		return nil, xerrors.Errorf("loading p2p key: %w", err)
+	}
+
+	h, ps, err := lp2p.ConstructHost(ds, priv, cfg.Addr, bootstrap)
+	if err != nil {
+		return nil, xerrors.Errorf("constructing host: %w", err)
+	}
+
+	addr, err := h.Network().InterfaceListenAddresses()
+	if err != nil {
+		return nil, xerrors.Errorf("getting InterfaceListenAddresses: %w", err)
+	}
+
+	for _, a := range addr {
+		log.Infof("%s/p2p/%s\n", a, h.ID())
+	}
+
+	t, err := ps.Join(lp2p.PubSubTopic(cfg.Network))
+	if err != nil {
+		return nil, xerrors.Errorf("joining topic: %w", err)
+	}
+
+	opts := []grpc.DialOption{}
+	if cfg.CertPath != "" {
+		creds, err := credentials.NewClientTLSFromFile(cfg.CertPath, "")
+		if err != nil {
+			return nil, xerrors.Errorf("loading cert file: %w", err)
+		}
+		opts = append(opts, grpc.WithTransportCredentials(creds))
+	} else if cfg.Insecure {
+		opts = append(opts, grpc.WithInsecure())
+	} else {
+		opts = append(opts, grpc.WithTransportCredentials(credentials.NewTLS(&tls.Config{})))
+	}
+	g := &GossipRelayNode{
+		bootstrap: bootstrap,
+		ds:        ds,
+		priv:      priv,
+		h:         h,
+		ps:        ps,
+		t:         t,
+		opts:      opts,
+		addr:      addr,
+		done:      make(chan struct{}),
+	}
+	go g.start(cfg.DrandPublicGRPC)
+	return g, nil
+}
+
+// Addr returns the gossipsub multiaddresses of this relay node.
+func (g *GossipRelayNode) Addr() []ma.Multiaddr {
```

Comment:
Minor point but as far as I can tell libp2p have specifically not used "addr" to refer to p2p addresses i.e. addresses with `/p2p/QmPeerId` suffix. Perhaps this should be renamed to `P2pAddrs()` or maybe `AddrInfo()` and return an [`peer.AddrInfo`](https://github.com/libp2p/go-libp2p-core/blob/master/peer/addrinfo.go). This is then trivially converted to a `[]multiaddr.Multiaddr` with [`peer.AddrInfoToP2pAddrs`](https://github.com/libp2p/go-libp2p-core/blob/264788628f5abdab1f08179e6a77964038217923/peer/addrinfo.go#L77-L91).

I think these can change over time (IDK maybe you're running a relay on your laptop) so it's perhaps better to get them from the libp2p host rather than cache them once when the host created.

Reply:
Renamed it to Multiaddrs() and getting them from Host.

If addresses can change over time, the notion of exporting them to give them to an external client doesn't make sense in the first place. Therefore, they are definitionally not expected to change. They can't change by themselves within Host, right?